lib/rubocop/cop/rspec/implicit_subject.rb

```diff
@@ -66,32 +94,70 @@ def on_send(node)
         private
 
         def autocorrect(corrector, node)
-          replacement = 'expect(subject)'
           case node.method_name
+          when :expect
+            corrector.replace(node, 'is_expected')
+          when :is_expected
+            corrector.replace(node.location.selector, 'expect(subject)')
           when :should
-            replacement += '.to'
+            corrector.replace(node.location.selector, 'expect(subject).to')
           when :should_not
-            replacement += '.not_to'
+            corrector.replace(node.location.selector, 'expect(subject).not_to')
           end
-
-          corrector.replace(node.loc.selector, replacement)
         end
 
-        def valid_usage?(node)
-          example = node.ancestors.find { |parent| example?(parent) }
-          return false if example.nil?
-
-          example.method?(:its) || allowed_by_style?(example)
+        def message(_node)
+          if style == :require_implicit
+            "Don't use explicit subject."
+          else
+            "Don't use implicit subject."
+          end
         end
 
-        def allowed_by_style?(example)
+        def invalid?(node)
           case style
+          when :require_implicit
+            invalid_to_require_implicit?(node)
+          when :disallow
+            invalid_to_disallow?(node)
           when :single_line_only
-            example.single_line?
+            invalid_to_single_line_only?(node)
           when :single_statement_only
-            !example.body.begin_type?
-          else
-            false
+            invalid_to_single_statement_only?(node)
+          end
+        end
+
+        def invalid_to_require_implicit?(node)
+          explicit_unnamed_subject?(node)
+        end
+
+        def invalid_to_disallow?(node)
+          implicit_subject?(node) && !its?(node)
+        end
+
+        def invalid_to_single_line_only?(node)
+          implicit_subject?(node) && !its?(node) && !single_line?(node)
+        end
+
+        def invalid_to_single_statement_only?(node)
+          implicit_subject?(node) && !its?(node) && !single_statement?(node)
+        end
```

Comment:
It requires a lot of "jumping" with the eyes to read this implementation. Would it be ok to have each of these one-line methods inlined into the `#invalid?` method?

Reply:
Unfortunately we can't do that because it violates both `Metrics/AbcSize` and `Metrics/CyclomaticComplexity` cops. This is why I splitted it into these methods.

Reply:
The only other alternative I can think of is
```rb
def invalid?(node)
  send("invalid_to_#{style}?", node)
end
``` 
which I don’t really like. So let’s keep it as it is then 😄 👍🏼